apollo-relay/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+    ///  - the reporter can't connect to the address
+    pub async fn try_new<T: AsRef<str>>(addr: T) -> Result<Self, ReporterError>
+    where
+        prost::bytes::Bytes: From<T>,
+    {
+        let ep = Endpoint::from_shared(addr)?;
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new_with_static(addr: &'static str) -> Result<Self, ReporterError> {
+        let ep = Endpoint::from_static(addr);
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Submit these stats onto the relayer for eventual processing.
+    ///
+    /// The relayer will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_stats(
+        &mut self,
+        graph: ReporterGraph,
+        q: String,
+        stats: ContextualizedStats,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_stats(Request::new(ReporterStats {
+                graph: Some(graph),
+                key: q,
+                stats: Some(stats),
+            }))
+            .await
+    }
+
+    /// Submit this trace onto the relayer for eventual processing.
+    ///
+    /// The relayer will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_trace(
+        &mut self,
+        graph: ReporterGraph,
+        q: String,
+        trace: Trace,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_trace(Request::new(ReporterTrace {
+                graph: Some(graph),
+                key: q,
+                trace: Some(trace),
+            }))
+            .await
+    }
+}
+
+/// The relay module contains the relaying components
+pub mod relay {
+    use super::report;
+    use crate::{ReporterStats, ReporterTrace, TracesAndStats};
+    use bytes::BytesMut;
+    use flate2::write::GzEncoder;
+    use flate2::Compression;
+    use prost::Message;
+    use prost_types::Timestamp;
+    use reqwest::Client;
+    use serde::{Deserialize, Serialize};
+    use std::collections::HashMap;
+    use std::io::Write;
+    use std::net::SocketAddr;
+    use std::sync::Arc;
+    use std::time::{SystemTime, UNIX_EPOCH};
+    use tokio::sync::mpsc::Sender;
+    use tokio::sync::Mutex;
+    use tokio::time::{interval, Duration, MissedTickBehavior};
+    use tonic::transport::{Error, Server};
+    use tonic::{Request, Response, Status};
+
+    pub use crate::agent::reporter_server::{Reporter, ReporterServer};
+    use crate::agent::{ReporterGraph, ReporterResponse};
+
+    /// Allows common transfer code to be more easily represented
+    #[allow(clippy::large_enum_variant)]
+    enum StatsOrTrace {
```

Comment:
Note that as of recently, the TracesAndStats message has a third thing: referenced_fields_by_type. This is based on extracting the referenced fields from the operation, and supports the new functionality on the Fields page where you can see what fields are *referenced* by an operation, not just those that are executed. See https://www.apollographql.com/docs/studio/metrics/field-usage/#field-executions-vs-referencing-operations

(For a given operation and schema this is always the same so figure out a reasonable place to cache that calculation. Of course we should eventually improve our reporting protocol so that we don't need to send it at all if Apollo's servers already know this op/schema pair...)

I do wonder if it might make sense to not divide the codepaths into traces and stats separately and to just act on the TracesAndStats object...